### PR TITLE
Flexible minting module 

### PR DESCRIPTION
**Pull request by BrandonWeng**
Allows users to control the minting schedule and amount based on a governance parameter that can be changed through a gov proposal. Takes a list of {date, token_release_amount}. The precision is up to 1 minute late in the first epoch of `date`

See that it's still able to mint new tokens

In the initial Genesis file:
```
    "mint": {
      "minter": {
        "epoch_provisions": "0.000000000000000000"
      },
      "params": {
        "mint_denom": "usei",
        "genesis_epoch_provisions": "5000000.000000000000000000",
        "token_release_schedule": [
            {
                "date": "2022-09-02",
                "token_release_amount": "2500000"
            },
            {
                "date": "2022-09-05",
                "token_release_amount": "1250000"
            },
            {
                "date": "2022-09-06",
                "token_release_amount": "625000"
            }
        ]
      }
    },
...

 ~/sei/brando  seid query bank total                                                            ✓  11618  01:26:09 
pagination:
  next_key: null
  total: "0"
supply:
- amount: "200000000000000000000"
  denom: usei
...
 ~/sei/brando  seid query bank total                                ✓  11631  01:34:24 
pagination:
  next_key: null
  total: "0"
supply:
- amount: "200000000000001250000"
  denom: usei
```

Submitting a gov proposal to update the param 
```
 ~/sei/brando  cat ./mint-param.json                                             
{
        "title": "Param Change for Bweng Testing",
        "description": "Changing minting schedule",
        "changes": [{
                "subspace": "mint",
                "key": "TokenReleaseSchedule",
                "value": [{
                        "date": "2022-09-01",
                        "token_release_amount": "50000"
                }, {
                        "date": "2023-09-01",
                        "token_release_amount": "25000"
                }, {
                        "date": "2025-09-01",
                        "token_release_amount": "123123"
                }]
        }],
         "deposit": "10000000usei"
}
...
seid tx gov submit-proposal param-change  ./mint-param.json  --from brando --chain-id sei --fees 20000usei
...
seid tx gov vote 1 yes --from brando --chain-id sei --fees 20000usei -y
....
```

In another terminal:
```
 ~/sei/brando  seid query params subspace mint TokenReleaseSchedule                
key: TokenReleaseSchedule
subspace: mint
value: '[{"date":"2022-09-01","token_release_amount":"2500000"},{"date":"2022-09-01","token_release_amount":"1250000"},{"date":"2025-09-01","token_release_amount":"625000"}]'
 ~/sei/brando  seid query params subspace mint TokenReleaseSchedule                
key: TokenReleaseSchedule
subspace: mint
value: '[{"date":"2022-09-01","token_release_amount":"50000"},{"date":"2023-09-01","token_release_amount":"25000"},{"date":"2025-09-01","token_release_amount":"123123"}]'
```

### CODE REVIEW

File: x/mint/types/minter.go [hunk 1]
```diff
@@ -30,13 +33,29 @@ func ValidateMinter(minter Minter) error {
...
 func (m Minter) EpochProvision(params Params) sdk.Coin {
 	provisionAmt := m.EpochProvisions
 	return sdk.NewCoin(params.MintDenom, provisionAmt.TruncateInt())
 }
+
+/*	Returns ScheduledRelease if the date of the block matches the scheduled release date.
+ *	You may only schedule one release of tokens on each day, the date must be in
+ *  types.TokenReleaseDateFormat.
+ */
+func GetScheduledTokenRelease(
```
Comment (BrandonWeng) [inline x/mint/types/minter.go:47]: Somewhat inefficient as we need to loop through the entire list of release dates, but the list will not grow beyond 10-20 years the list should be bounded by O(10-20). 

Happy to hear any suggestions for how to make this more efficient, one idea I had was to store the index of the previously applied list in the chain but that would be very fragile to param-changes in the future

File: x/mint/keeper/hooks.go [hunk 2]
```diff
@@ -14,18 +14,20 @@ func (k Keeper) BeforeEpochStart(ctx sdk.Context, epoch epochTypes.Epoch) {
...
+	scheduledTokenRelease := types.GetScheduledTokenRelease(epoch, lastRelaseDate, params.GetTokenReleaseSchedule())
+	if scheduledTokenRelease == nil {
+		ctx.Logger().Debug(fmt.Sprintf("No release at epoch time %s; last release %s", epoch.GetCurrentEpochStartTime().String(), lastRelaseDate.Format(types.TokenReleaseDateFormat)))
 		return
 	}
-	// Halven the reward per halven period
-	minter := k.GetMinter(ctx)
-	minter.EpochProvisions = minter.NextEpochProvisions(params)
+
+	minter.EpochProvisions = sdk.NewDec(scheduledTokenRelease.GetTokenReleaseAmount())
```
Comment (BrandonWeng) [inline x/mint/keeper/hooks.go:27]: Still keeping `EpochProvisions` around as a way to tell the amount of tokens that was last minted

File: x/mint/types/minter.go (hunk 1 above)
Comment (philipsu522) [reply x/mint/types/minter.go:47]: This is fine, let's keep it as is

File: x/mint/types/params.go [hunk 3]
```diff
@@ -13,35 +14,32 @@ import (
...
-	KeyGenesisEpochProvisions  = []byte("GenesisEpochProvisions")
-	KeyReductionPeriodInEpochs = []byte("ReductionPeriodInEpochs")
-	KeyReductionFactor         = []byte("ReductionFactor")
+	KeyMintDenom              = []byte("MintDenom")
+	KeyGenesisEpochProvisions = []byte("GenesisEpochProvisions")
+	KeyTokenReleaseSchedule   = []byte("TokenReleaseSchedule")
 )
 
-// ParamTable for minting module.
+// ParamTable for minting m1odule.
```
Comment (philipsu522) [inline x/mint/types/params.go]: nit - typo

File: x/mint/types/params.go [hunk 4]
```diff
@@ -107,32 +101,19 @@ func validateGenesisEpochProvisions(i interface{}) error {
...
-		return fmt.Errorf("max validators must be positive: %d", v)
-	}
-
-	return nil
-}
-
-func validateReductionFactor(i interface{}) error {
-	v, ok := i.(sdk.Dec)
+func validateTokenReleaseSchedule(i interface{}) error {
+	tokenReleaseSchedule, ok := i.([]ScheduledTokenRelease)
```
Comment (philipsu522) [inline x/mint/types/params.go:105]: can we also check that date is in the right "yyyy-mm-dd" format, and other basic things like:
- mm, dd > 0
- yyyy > 1970
- mm < 12
- dd <= 31
- dd logic:
- - if mm == 1,3,5,7,8,10,12, dd == 31
- - if mm == 4,6,9,11, dd ==30
- - if mm == 2 and yy % 4 != 0, dd == 28
- - if mm == 2 and yy % $ == 0, dd == 29
Re. leap year, seems like the logic is a bit more complicated:
```
Years that are divisible by 100 (century years such as 1900 or 2000) cannot be leap years unless they are also divisible by 400. (For this reason, the years 1700, 1800, and 1900 were not leap years, but the years 1600 and 2000 were.)
```
but think we can ignore this, maybe just make a note

Comment (BrandonWeng) [reply x/mint/types/params.go:105]: Hmm, I was hoping that the DateTime parser should be catching almost all the cases you mentioned above 

`TokenReleaseDateFormat == yyyy-mm-dd` here
```
		_, err := time.Parse(TokenReleaseDateFormat, scheduledTokenRelease.GetDate())
		if err != nil {
			return fmt.Errorf("error: invalid date format use yyyy-mm-dd: %s", err)
		}
```

Comment (BrandonWeng) [reply x/mint/types/params.go:105]: e.g 
image

Comment (BrandonWeng) [reply x/mint/types/params.go:105]: one thing it's not checking is if the year is > 1970 but if it is we won't be hitting that condition anyways

**Review by philipsu522 — APPROVED** [review]

**Review by LCyson — APPROVED** [review]

**Pull request merged**